packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx

```diff
@@ -0,0 +1,233 @@
+  }, [depth, subNavigation]);
+
+  const handleSidebarItemClick = React.useCallback(
+    (itemId: string) => () => {
+      setExpandedSidebarItemIds((previousValue) =>
+        previousValue.includes(itemId)
+          ? previousValue.filter((previousValueItemId) => previousValueItemId !== itemId)
+          : [...previousValue, itemId],
+      );
+    },
+    [],
+  );
+
+  return (
+    <List sx={{ mb: depth === 0 ? 4 : 1, pl: 2 * depth }}>
+      {subNavigation.map((navigationItem, navigationItemIndex) => {
+        if (navigationItem.kind === 'header') {
+          return (
+            <ListSubheader key={`subheader-${depth}-${navigationItemIndex}`} component="div">
+              {navigationItem.title}
+            </ListSubheader>
+          );
+        }
+
+        if (navigationItem.kind === 'divider') {
+          const nextItem = subNavigation[navigationItemIndex + 1];
+
+          return (
+            <Divider
+              key={`divider-${depth}-${navigationItemIndex}`}
+              sx={{ mt: 1, mb: nextItem.kind === 'header' ? 0 : 1 }}
+            />
+          );
+        }
+
+        const navigationItemId = `${navigationItem.title}-${depth}-${navigationItemIndex}`;
+
+        const isNestedNavigationExpanded = expandedSidebarItemIds.includes(navigationItemId);
+
+        const nestedNavigationCollapseIcon = isNestedNavigationExpanded ? (
+          <ExpandLessIcon />
+        ) : (
+          <ExpandMoreIcon />
+        );
+
+        const listItem = (
+          <ListItem>
+            <ListItemButton
+              selected={IS_CLIENT && hasMounted && window.location.pathname === navigationItem.path}
+              onClick={handleSidebarItemClick(navigationItemId)}
+            >
+              <ListItemIcon>{navigationItem.icon}</ListItemIcon>
+              <ListItemText primary={navigationItem.title} />
+              {navigationItem.children ? nestedNavigationCollapseIcon : null}
+            </ListItemButton>
+          </ListItem>
+        );
+
+        return (
+          <React.Fragment key={navigationItemId}>
+            {navigationItem.path ? (
+              <a href={navigationItem.path} style={{ color: 'inherit', textDecoration: 'none' }}>
+                {listItem}
+              </a>
+            ) : (
+              listItem
+            )}
+            {IS_CLIENT && hasMounted && navigationItem.children ? (
+              <Collapse in={isNestedNavigationExpanded} timeout="auto" unmountOnExit>
+                <DashboardSidebarSubNavigation
+                  subNavigation={navigationItem.children}
+                  depth={depth + 1}
+                />
+              </Collapse>
+            ) : null}
+          </React.Fragment>
+        );
+      })}
+    </List>
+  );
+}
+
+interface DashboardLayoutProps {
+  children: React.ReactNode;
+  sidebarWidth?: number;
+  hideTitle?: boolean;
+}
+
+export function DashboardLayout({
+  children,
+  sidebarWidth = DEFAULT_DRAWER_WIDTH,
+  hideTitle = false,
```

Comment:
I feel like at this point it's hard to say whether this behavior should be enabled through a specific prop, or we should let it be solved by slots API.
Perhaps at this stage we should concentrate more on coming up with the best possible default experience, for all breakpoints and think of customization once we've collect a few use cases.